pkg_install-testing/files/info/pkg_info.1
.\" $NetBSD: pkg_info.1,v 1.2 2004/05/22 19:56:26 jeremy-c-reed Exp $
.\"
.\" FreeBSD install - a package for the installation and maintenance
.\" of non-core utilities.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" Jordan K. Hubbard
.\"
.\"
.\"     @(#)pkg_info.1
.\"
.Dd January 6, 2004
.Dt PKG_INFO 1
.Os
.Sh NAME
.Nm pkg_info
.Nd a utility for displaying information on software packages
.Sh SYNOPSIS
.Nm
.Op Fl BbcDdFfhIikLmNnpqRrSsVv
.Bk -words
.Op Fl e Ar package
.Ek
.Bk -words
.Op Fl K Ar pkg_dbdir
.Ek
.Bk -words
.Op Fl l Ar prefix
.Ek
.Ar pkg-name ...
.Nm
.Bk -words
.Op Fl a Ar flags
.Ek
.Nm
.Bk -words
.Op Fl Q Ar variable
.Ek
.Ar pkg-name ...
.Sh DESCRIPTION
The
.Nm
command is used to dump out information for packages, which may be either
packed up in files or already installed on the system with the
.Xr pkg_create 1
command.
.Pp
The
.Ar pkg-name
may be the name of an installed package (with our without version), a
pattern matching several installed packages (see the
.Sx PACKAGE WILDCARDS
section for a description of possible patterns),
the pathname to a
binary package, a filename belonging to an installed
package (if
.Fl F
is also given), or a URL to an ftp-available package.
.Pp
The following command-line options are supported:
.Bl -tag -width indent
.It Fl a
Show information for all currently installed packages.
.It Fl B
Show some of the important definitions used when building
the binary package (the "Build information") for each package.
.It Fl b
Show the
.Nx
RCS Id strings from the files used in the construction
of the binary package (the "Build version") for each package.
These files are the package Makefile, any patch files, any checksum
files, and the packing list file.
.It Fl c
Show the one-line comment field for each package.
.It Fl D
Show the install-message file (if any) for each package.
.It Fl d
Show the long-description field for each package.
.It Fl e Ar pkg-name
This option
allows you to test for the existence of a given package.
If the package identified by
.Ar pkg-name
is currently installed, return code is 0, otherwise 1.
The names of any package(s) found installed are printed to
stdout unless turned off using the
.Fl q
option.
.Ar pkg-name
can contain wildcards, see the
.Sx PACKAGE WILDCARDS
section below.
.It Fl F
Interpret any pkg-name given as filename, and translate it to a
package name using the Package Database.
This can be used to query information on a per-file basis, e.g. in
conjunction with the
.Fl e
flag to find out which package a file belongs to.
.It Fl f
Show the packing list instructions for each package.
.It Fl I
Show the index entry for each package.
.It Fl i
Show the install script (if any) for each package.
.It Fl K Ar pkg_dbdir
Set
.Ar pkg_dbdir
as the package database directory.
If this option isn't specified, then the package database directory is
taken from the value of the environment variable
.Ev PKG_DBDIR
if it's set, otherwise it defaults to
.Pa /var/db/pkg .
.It Fl k
Show the de-install script (if any) for each package.
.It Fl L
Show the files within each package.
This is different from just viewing the packing list, since full pathnames
for everything are generated.
.It Fl l Ar str
Prefix each information category header (see
.Fl q )
shown with
.Ar str .
This is primarily of use to front-end programs that want to request a
lot of different information fields at once for a package, but don't
necessary want the output intermingled in such a way that they can't
organize it.
This lets you add a special token to the start of each field.
.It Fl m
Show the mtree file (if any) for each package.
.It Fl N
Show which packages each package was built with (exact dependencies), if any.
.It Fl n
Show which packages each package needs (depends upon), if any.
.It Fl p
Show the installation prefix for each package.
.It Fl Q
Show the definition of
.Ar variable
from the build information for each package.
An empty string is returned if no such variable definition is found for
the package(s).
.It Fl q
Be ``quiet'' in emitting report headers and such, just dump the
raw info (basically, assume a non-human reading).
.It Fl R
Show which packages are required by each package.
.It Fl r
Show the requirements script (if any) for each package.
.It Fl S
Show the size of this package and all the packages it requires,
in bytes.
.It Fl s
Show the size of this package in bytes.
The size is calculated by adding up the size of each file of the package.
.It Fl V
Print version number and exit.
.It Fl v
Turn on verbose output.
.El
.Sh TECHNICAL DETAILS
Package info is either extracted from package files named on the
command line, or from already installed package information
in
.Pa /var/db/pkg/\*[Lt]pkg-name\*[Gt] .
.Pp
A filename can be given instead of a (installed) package name to query
information on the package this file belongs to.
This filename is then resolved to a package name using the Package Database.
For this translation to take place, the
.Fl F
flag must be given.
The filename must be absolute, compare the output of pkg_info
.Fl aF .
.Sh PACKAGE WILDCARDS
In the places where a package name/version is expected, e.g. for the
.Fl e
switch, several forms can be used.
Either use a package name with or without version, or specify a
package wildcard that gets matched against all installed packages.
.Pp
Package wildcards use
.Xr fnmatch 3 .
In addition,
.Xr csh 1
style {,} alternates have also been implemented.
Package version numbers can also be matched in a relational manner
using the
.Pa \*[Ge] , \*[Le] , \*[Gt] ,
and
.Pa \*[Lt]
operators.
For example,
.Pa pkg_info -e 'name\*[Ge]1.3'
will match versions 1.3 and later of the
.Pa name
package.
The collating sequence of the various package version numbers is
unusual, but strives to be consistent.
The magic string ``alpha'' equates to
.Pa alpha version
and sorts before a beta version.
The magic string ``beta'' equates to
.Pa beta version
and sorts before a release candidate.
The magic string ``rc'' equates to
.Pa release candidate
and sorts before a release.
For example,
.Pa name-1.3rc3
will sort before
.Pa name-1.3
and after
.Pa name-1.2.9
Similarly
.Pa name-1.3alpha2
will sort before
.Pa name-1.3beta1
and they both sort before
.Pa name-1.3rc1
In addition, alphabetic characters sort in the same place as
their numeric counterparts, so that
.Pa name-1.2e
has the same sorting value as
.Pa name-1.2.5
The magic string ``pl'' equates to a
.Pa patch level
and has the same value as a dot in the dewey-decimal ordering schemes.
.Sh ENVIRONMENT
.Bl -tag -width PKG_DBDIR
.It Ev PKG_DBDIR
If the
.Fl K
flag isn't given, then
.Ev PKG_DBDIR
is the location of the package database directory.
The default package database directory is
.Pa /var/db/pkg .
.It Ev PKG_PATH
This can be used to specify a semicolon-separated list of paths and URLs to search for
package files.
If
.Ev PKG_PATH
is used, the suffix
.Pa .tgz
is automatically appended to the
.Ar pkg-name ,
whereas searching in the current directory uses
.Ar pkg-name
literally.
.It Ev PKG_TMPDIR , Ev TMPDIR
These are tried in turn (if set) as candidate directories in which
to create a ``staging area'' for any files extracted by
.Nm
from package files.
If neither
.Ev PKG_TMPDIR
nor
.Ev TMPDIR
yields a suitable scratch directory,
.Pa /var/tmp ,
.Pa /tmp ,
and
.Pa /usr/tmp
are tried in turn.
Note that
.Pa /usr/tmp
may be created, if it doesn't already exist.
.Pp
Since
.Nm
requires very little information to be extracted from any package
files examined, it is unlikely that these environment variables would
ever need to be used to work around limited available space in the
default locations.
.El
.Sh SEE ALSO
.Xr pkg_add 1 ,
.Xr pkg_admin 1 ,
.Xr pkg_create 1 ,
.Xr pkg_delete 1 ,
.Xr mktemp 3 ,
.Xr packages 7 ,
.Xr mtree 8
.Sh AUTHORS
.Bl -tag -width indent -compact
.It "Jordan Hubbard"
most of the work
.It "John Kohl"
refined it for
.Nx
.It "Hubert Feyrer"
.Nx
wildcard dependency processing, pkgdb, depends displaying,
pkg size display etc.
.El


hepmc/ChangeLog
2010-08-18  Lynn Garren

      * src/GenEventStreamIO.cc, StreamHelpers.cc: 
        cleanup unassociated event remnants if corrupt data is found

2010-07-30  Lynn Garren & Andy Buckley

      * src/GenEventStreamIO.cc: create GenCrossSection, PdfInfo, and HeavyIon
        on the stack rather than the heap to fix a memory leak

  --------------------------  HepMC-2.06.01  --------------------------
2010-06-01  Lynn Garren

      * HepMC/GenRanges.h: remove ConstGenVertexParticleRange since 
        there is neither a GenVertex::particle_const_iterator nor a 
        GenVertex::vertex_const_iterator 
      * HepMC/GenVertex.h: remove unimplemented 
        ConstGenVertexParticleRange particles()


  --------------------------  HepMC-2.06.00  --------------------------
  --------------------------  HepMC-2.06.00.b02  --------------------------
2010-05-13  Lynn Garren

      * HepMC/Version.h: version() now takes an optional output stream

      * HepMC/GenEvent.h: add helper method for reading weight names
      * src/GenEventStreamIO.cc: use new helper method to read weight names
      * testHepMCVarious.input: change order of some event info lines
      * test/testIOGenEvent.input: add named weights for testing

2010-05-12  Lynn Garren

      * "make check" now works if momentum units are MEV

2010-05-11  Lynn Garren

      * rename HepMC/GenEventIterators.h to HepMC/GenRanges.h
      
      * src/GenEventStreamIO.cc: fix a bug reading unnamed (old style) weights
      
      * test/testHepMC: add more weight testing

  --------------------------  HepMC-2.06.00.b01  --------------------------

2010-04-05  Lynn Garren

      * HepMC/StreamInfo.h: add reading_event_header flag for use by streaming input
      * src/GenEventStreamIO.cc: Input no longer expects a fixed order of 
        header lines and will ignore lines that are not of a known type.

      * HepMC/HerwigWrapper.h: add hwbeam
      * rename PythiaWrapper6_2.h to PythiaWrapper6_4.h 
        Add structs for more common blocks.

2010-04-02  Lynn Garren

      * HepMC/IteratorRange.h: put the enum in its own header
      * HepMC/GenEvent.h: add vertex_range() and particle_range()
      * HepMC/GenParticle.h: add particles_in(range) and particles_out(range)
      * HepMC/GenVertex.h: add particles(range), 
        particles_in(GenParticle,range), and particles_out(GenParticle,range)
      * test/testHepMCIteration.cc: use the new methods

2010-03-20  Lynn Garren

      * HepMC/GenEventIterators.h: add ConstGenVertexParticleRange,
        ConstGenParticleProductionRange, and ConstGenParticleEndRange
      * HepMC/WeightContainer.h: method to check for weight name is has_key()
      * src/GenEventStreamIO.cc: use quotes to delineate weight names

2010-02-08  Lynn Garren

      * remove unused PythiaWrapper5_720.h, PythiaWrapper6_152.h, and PythiaWrapper6_152_WIN32.h
      * combine HerwigWrapper6_4.h and HerwigWrapper.h in a single file
      * HepMC/PythiaWrapper.h: add getPythiaCrossSection function
      * HepMC/HerwigWrapper.h: add getHerwigCrossSection function
      * fio/PythiaWrapper.cc: implement getPythiaCrossSection function
      * fio/HerwigWrapper.cc: implement getHerwigCrossSection function
      * remove redundant examples/HerwigHelper.h.

2010-02-04  Lynn Garren

      * doc/Makefile.am, examples/Makefile.am: use pkgdatadir to define the instation directory
        The default installation directory is now ${prefix}/share/HepMC/doc
        Override with --datadir=/some/directory/path

      * HepMC/Polarization.h: add is_defined() and set_undefined()
        is_defined returns false if the polarization is undefined
	set_undefined unsets the defined flag and zeroes the Polarization

2010-01-28  Lynn Garren

      * HepMC/GenEventIterators.h: This new header defines a set of classes
        which provide the appropriate iterators ranges.  They can be used to 
	simplify code and also directly with such utilities as Boost foreach.
	These new classes are 
	GenEventVertexRange, ConstGenEventVertexRange, 
	GenEventParticleRange, ConstGenEventParticleRange, 
	GenVertexParticleRange, 
	GenParticleProductionRange, and GenParticleEndRange
      * test/testHepMCIteration.cc: use the new iterator clases and
        also illustrate the use of std::for_each
 
2010-01-26  Lynn Garren

      * test/testHepMCIteration: illustrate use of std::for_each
 
2010-01-25  Lynn Garren

      * HepMC/WeightContainer.h: add map-like functionality, allowing users to
        name weights, while retaining the original vector interface for 
	complete backwards compatibility
	If a name is not specified for a particular weight, the default
	name is the weight's vector index.
	Also add == and != operators.
      * HepMC/GenEvent.h: add a private method used by the stream I/O
      * src/GenEventStreamIO.cc: add a new line for weight names
        The weight name line is optional when reading input and is 
	written only if the event weights object is not empty.
      * test/testWeights.cc: add tests for WeightContainer
      * test/testHepMC: test I/O of named weights
      * examples/testPythiaCopies.cc: set some arbitrary weights
 
2010-01-09  Lynn Garren

      * removed ParticleDataTable.h and all supporting references.

 --------------------------  HepMC-2.05.01  --------------------------

2010-01-04  Lynn Garren

      * HepMC/GenEvent.h: add write_cross_section(std::ostream&)
      * src/GenEvent.cc: output the cross section (if it exists) when printing an event
      * src/GenVertex.cc, GenParticle.cc: save and restore output stream state
        when calling print methods
	
      * examples: Calculate Pythia and Herwig cross sections using
        getPythiaCrossSection from examples/PythiaHelper.h and 
	getHerwigCrossSection from examples/HerwigHelper.h.
        Set the units explicitly.
      * test/testHepMC.cc.in: call write_cross_section()

2009-12-28  Lynn Garren

       * examples/VectorConversion.h: inline the functions


pkg_select/files/pkg_info.c
/*
 * Copyright (c) 2005
 *      iMil <imil@gcu.info>.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *      This product includes software developed by iMil.
 * 4. Neither the name of the author nor the names of any co-contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY iMil AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL iMil OR THE VOICES IN HIS HEAD
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
 * THE POSSIBILITY OF SUCH DAMAGE.
 *
 * $Id: pkg_info.c,v 1.1.1.1 2005/02/21 10:38:07 imilh Exp $ 
 */

#include "pkg_select.h"

static int show_pkgfile(WINDOW *, char *, const char *);
static void pkg_popup(WINDOW *, char *, char *, 
		      const char *, const char *, const char *);
static void pkg_depends(WINDOW *, char *, const char *);

static int
show_pkgfile(WINDOW *win, char *path, const char *file)
{
	char buf[MAXLEN];

	snprintf(buf, MAXLEN, "%s/%s", path, file);
	/* if 'q' was pressed, return 0 so we quit */
	return(more_file(win, buf, LINES - 2, COLS - 2, 1, 1));
}

static void
pkg_popup(WINDOW *win, char *path, char *pkg, 
	  const char *label, const char *msg, const char *action)
{
	int len;
	char rep, *env, pkg_path[MAXLEN];

	len = strlen(msg) + 4;
	rep = getch_popup(label, msg, 
			  5, len, 
			  (LINES / 2) - 2, (COLS / 2) - (len / 2));

	if (rep != 'n') {
		wclear(win);
		wrefresh(win);

		/* recursive pkg_delete was asked */
		if (*action == 'd' && rep == 'r')
			PKG_TOOL("delete", pkg, "-r");

		/* make action */
		if ((*action == 'i' || *action == 'u' || *action == 'd') && 
		    (rep != 'r'))
				PKGSRC_MAKE(action, path);

		/* pkg_* action */
		if (*action == 'a') {
			env = getenv("PKG_PATH");
			if (env == NULL) {
				build_pkg_path(&pkg_path[0]);
				mvwprintw(win, 0, 0, 
					"PKG_PATH variable not set, default to\n%s\n", pkg_path);
				setenv("PKG_PATH", pkg_path, 1);
				wprintw(win, ANY_KEY);
				(void) wgetch(win);

				wclear(win);
				wrefresh(win);
			}
			PKG_TOOL(action, "", pkg);
		} /* answer was yes */

		wclear(win);
		wrefresh(win);
	}
}

static void
pkg_depends(WINDOW *win, char *path, const char *varname)
{
	int len;
	char **list;
	WINDOW *popup;
	char buf[MAXLEN];

	snprintf(buf, MAXLEN, "%s %s", PLEASE_WAIT_CALC, varname);
	len = strlen(buf) + 4;
	
	popup = info_popup("information", buf, 
			   5, len, 
			  (LINES / 2) - 2, (COLS / 2) - (len / 2));
	wrefresh(popup);

	list = show_var(path, varname, T_TRUE);
	/* enter main_loop */
	nodir_loop(path, list);

	free_list(&list);
	delwin(popup);
}

int
info_win(WINDOW *win, char *pkg, char *path)
{
	int ret, c, line_index, msv, len, recurs;
	struct cf *file;
	char buf[MAXLEN], *homepage, *comment, *distname, *version;
	WINDOW *popup;

	snprintf(buf, MAXLEN, "%s/MESSAGE", path);

	/* info window needs all screen */
	wresize(win, LINES - 2, COLS - 2);
	
	/* deep recursion check */
	if (strstr(path, "../") != NULL)
		recurs = T_TRUE;
	else
		recurs = T_FALSE;

	for (;;) {
		/* load Makefile* */
		file = load_makefile(path, FULL);

		line_index = 2;

		wclear(win);
		box(win, 0, 0);
		wattron(win, A_BOLD);

		mvwprintw(win, 0, COLS - 8 - strlen(path), "[ %s ]", path);

		mvwprintw(win, line_index, 2, PKG_ADMIN_INFO);
		wattroff(win, A_BOLD);
		line_index += 2;

		/* get values from Makefile slist */
		distname = getval(file, "DISTNAME=");
		homepage = getval(file, "HOMEPAGE=");
		comment = getcomment(file);
		
		msv = 0;
		/* one or more members could not be read from Makefile 
		 * use make show-var to fill all of them
		 */
		if (distname == NULL || comment == NULL) {
			len = strlen(SEARCH_INFOS) + 4;

			/* this cound take a moment, warn user */
			popup = info_popup("information", 
					   SEARCH_INFOS, 
					   5, len, 
					   (LINES / 2) - 2, 
					   (COLS / 2) - (len / 2));
			wrefresh(popup);

			distname = show_first_var(path, "DISTNAME");
			homepage = show_first_var(path, "HOMEPAGE");
			comment = show_first_var(path, "COMMENT");
			/* set make show-var variable */
			msv = 1;

			delwin(popup);
		}

		if (distname == NULL) {
			wprint_kb(win, "package:", "none", line_index, 2);
		} else
			wprint_kb(win, "distname:", distname, line_index, 2);
		line_index++;

		if ((version = getpkginfo(pkg, PKG_VERSION)) != NULL)
			wprint_kb(win, "installed:", version, line_index, 2);
		else
			wprint_kb(win, "installed:", "none", line_index, 2);
		line_index++;

		if (homepage == NULL) {
			wprint_kb(win, "homepage:", "none", line_index, 2);
		} else
			wprint_kb(win, "homepage:", homepage, line_index, 2);
		line_index++;

		if (comment == NULL)
			wprint_kb(win, "comment:", "none", line_index, 2);
		else {
			/* 7 == strlen("comment:") + 2 borders + 
			   "..." + 2 spaces */
			cut_str(comment, COLS - 17);
			wprint_kb(win, "comment:", comment, line_index, 2);
		}
		line_index += 2;
		
		wprint_kb(win, "[m]", "show package description", 
			  line_index, 2);
		line_index++;

		if (file_exists(buf)) {
			wprint_kb(win, "[v]", 
				  "view package message", line_index, 2);
			line_index++;
		}

		if (!recurs) {
			wprint_kb(win, "[e]", "show package dependancies",
				  line_index, 2);
			line_index++;
		}
		line_index++;

		wprint_kb(win, "[i]", "build and install package",
			  line_index, 2);
		line_index++;
		wprint_kb(win, "[u]", "build and upgrade package",
			  line_index, 2);
		line_index++;

		wprint_kb(win, "[b]", "install binary package", line_index, 2);
		line_index++;

		wprint_kb(win, "[d]", "de-install package", line_index, 2);

		wprint_kb(win, "[<-]", "back", LINES - BOTTOM_KB, 2);
		wprint_kb(win, "[q]", "quit", LINES - BOTTOM_KB, 14);
	
		/* we got infos from make show-var, must free */
		if (msv) {
			XFREE(distname);
			XFREE(homepage);
			XFREE(comment);
		}
	
		c = wgetch(win);
		switch (c) {
		case 'm':
			ret = show_pkgfile(win, path, "DESCR");
			if (!ret)
				/* quit was asked */
				return(ret);
			break;
		case 'v':
			ret = show_pkgfile(win, path, "MESSAGE");
			if (!ret)
				/* quit was asked */
				return(ret);
			break;
		case 'i':
			pkg_popup(win, path, pkg, 
				  "Install package", 
				  "build and install package ? [Y/n]",
				  "install");
			break;
		case 'u':
			pkg_popup(win, path, pkg, 
				  "Upgrade package", 
				  "build and upgrade package ? [Y/n]",
				  "update");
			break;
		case 'd':
			pkg_popup(win, path, pkg, 
				  "De-install package", 
				  "de-install package ? [Y/n/r(ecursive)]",
				  "deinstall");
			break;
		case 'b':
			pkg_popup(win, path, pkg, 
				  "Install package",
				  "install binary package ? [Y/n]",
				  "add");
			break;
		case 'e':
			if (!recurs)
				pkg_depends(win, path, "DEPENDS");
			/* back from main_loop, resize */
			wresize(win, LINES - 2, COLS - 2);

			break;
		case 'q':
			/* if we are on a depend list, we want a clean exit */
			freecf(file);
			return(IN_QUIT);

			break;
		case KEY_LEFT:
			freecf(file);
			return(1);
			break;
		}

	}
	/* NOTREACHED */
	return(1);
}
